### ISSUE TITLE
No Serial output from Nucleo F401RE when trying to get data from SATEL VL53L8

### ISSUE THREAD

**Issue by yaswanthsk04** [stm32duino/VL53L8CX#8]
**Describe the bug**
I am trying to run VL53L8CX_HelloWorld_I2C.ino on Nucleo-F401RE with satel VL53L8 breakout board. But no output in serial monitor.

the program hangs in the init function.
![WhatsApp Image 2024-08-04 at 15 06 06_65f69d4c](https://github.com/user-attachments/assets/dac2bbfd-2c9d-45aa-9fe0-f5df10ffe5d5)

**To Reproduce**

1). Connections as per instructions from Github repo
2). Installed latest stm32duino boards
3). Installed Library from arduino Lib manager
4). Compile and Build 
5). Open serial Monitor with baud rate 460800
6). No output in serial monitor

**Expected behavior**
Expecting Output of sensor data in serial monitor

**Code Used to debug:**
```
#include <vl53l8cx.h>


#ifdef ARDUINO_SAM_DUE
  #define DEV_I2C Wire1
#else
  #define DEV_I2C Wire
#endif
#define SerialPort Serial

#define LPN_PIN A3
#define PWREN_PIN 11

void print_result(VL53L8CX_ResultsData *Result);
void clear_screen(void);
void handle_cmd(uint8_t cmd);
void display_commands_banner(void);

// Components.
VL53L8CX sensor_vl53l8cx_top(&DEV_I2C, LPN_PIN);

bool EnableAmbient = false;
bool EnableSignal = false;
uint8_t res = VL53L8CX_RESOLUTION_4X4;
char report[256];
uint8_t status;

/* Setup ---------------------------------------------------------------------*/
void setup()
{
 SerialPort.begin(460800);
  SerialPort.println("Entered Setup");
  // Enable PWREN pin if present
  if (PWREN_PIN >= 0) {
    pinMode(PWREN_PIN, OUTPUT);
    digitalWrite(PWREN_PIN, HIGH);
    delay(10);
  }
  SerialPort.println("Finished Setting up PWREN");
  

  // Initialize I2C bus.
  DEV_I2C.begin();
  SerialPort.println("Finished Setting up I2C Begin");
  // Configure VL53L8CX component.
  sensor_vl53l8cx_top.begin();
  SerialPort.println("Finished Setting up Sensor Begin");
  status = sensor_vl53l8cx_top.init();
  SerialPort.println("Finished Setting up Sensor Init");

  // Start Measurements
  status = sensor_vl53l8cx_top.start_ranging();
}

void loop()
{
  VL53L8CX_ResultsData Results;
  uint8_t NewDataReady = 0;

  do {
    status = sensor_vl53l8cx_top.check_data_ready(&NewDataReady);
  } while (!NewDataReady);

  if ((!status) && (NewDataReady != 0)) {
    status = sensor_vl53l8cx_top.get_ranging_data(&Results);
    print_result(&Results);
  }

  if (Serial.available() > 0) {
    handle_cmd(Serial.read());
  }
  delay(100);
}

void print_result(VL53L8CX_ResultsData *Result)
{
  int8_t i, j, k, l;
  uint8_t zones_per_line;
  uint8_t number_of_zones = res;

  zones_per_line = (number_of_zones == 16) ? 4 : 8;

  display_commands_banner();

  SerialPort.print("Cell Format :\n\n");

  for (l = 0; l < VL53L8CX_NB_TARGET_PER_ZONE; l++) {
    snprintf(report, sizeof(report), " \033[38;5;10m%20s\033[0m : %20s\n", "Distance [mm]", "Status");
    SerialPort.print(report);

    if (EnableAmbient || EnableSignal) {
      snprintf(report, sizeof(report), " %20s : %20s\n", "Signal [kcps/spad]", "Ambient [kcps/spad]");
      SerialPort.print(report);
    }
  }

  SerialPort.print("\n\n");

  for (j = 0; j < number_of_zones; j += zones_per_line) {
    for (i = 0; i < zones_per_line; i++) {
      SerialPort.print(" -----------------");
    }
    SerialPort.print("\n");

    for (i = 0; i < zones_per_line; i++) {
      SerialPort.print("|                 ");
    }
    SerialPort.print("|\n");

    for (l = 0; l < VL53L8CX_NB_TARGET_PER_ZONE; l++) {
      // Print distance and status
      for (k = (zones_per_line - 1); k >= 0; k--) {
        if (Result->nb_target_detected[j + k] > 0) {
          snprintf(report, sizeof(report), "| \033[38;5;10m%5ld\033[0m  :  %5ld ",
                   (long)Result->distance_mm[(VL53L8CX_NB_TARGET_PER_ZONE * (j + k)) + l],
                   (long)Result->target_status[(VL53L8CX_NB_TARGET_PER_ZONE * (j + k)) + l]);
          SerialPort.print(report);
        } else {
          snprintf(report, sizeof(report), "| %5s  :  %5s ", "X", "X");
          SerialPort.print(report);
        }
      }
      SerialPort.print("|\n");

      if (EnableAmbient || EnableSignal) {
        // Print Signal and Ambient
        for (k = (zones_per_line - 1); k >= 0; k--) {
          if (Result->nb_target_detected[j + k] > 0) {
            if (EnableSignal) {
              snprintf(report, sizeof(report), "| %5ld  :  ", (long)Result->signal_per_spad[(VL53L8CX_NB_TARGET_PER_ZONE * (j + k)) + l]);
              SerialPort.print(report);
            } else {
              snprintf(report, sizeof(report), "| %5s  :  ", "X");
              SerialPort.print(report);
            }
            if (EnableAmbient) {
              snprintf(report, sizeof(report), "%5ld ", (long)Result->ambient_per_spad[j + k]);
              SerialPort.print(report);
            } else {
              snprintf(report, sizeof(report), "%5s ", "X");
              SerialPort.print(report);
            }
          } else {
            snprintf(report, sizeof(report), "| %5s  :  %5s ", "X", "X");
            SerialPort.print(report);
          }
        }
        SerialPort.print("|\n");
      }
    }
  }
  for (i = 0; i < zones_per_line; i++) {
    SerialPort.print(" -----------------");
  }
  SerialPort.print("\n");
}

void toggle_resolution(void)
{
  status = sensor_vl53l8cx_top.stop_ranging();

  switch (res) {
    case VL53L8CX_RESOLUTION_4X4:
      res = VL53L8CX_RESOLUTION_8X8;
      break;

    case VL53L8CX_RESOLUTION_8X8:
      res = VL53L8CX_RESOLUTION_4X4;
      break;

    default:
      break;
  }
  status = sensor_vl53l8cx_top.set_resolution(res);
  status = sensor_vl53l8cx_top.start_ranging();
}

void toggle_signal_and_ambient(void)
{
  EnableAmbient = (EnableAmbient) ? false : true;
  EnableSignal = (EnableSignal) ? false : true;
}

void clear_screen(void)
{
  snprintf(report, sizeof(report), "%c[2J", 27); /* 27 is ESC command */
  SerialPort.print(report);
}

void display_commands_banner(void)
{
  snprintf(report, sizeof(report), "%c[2H", 27); /* 27 is ESC command */
  SerialPort.print(report);

  Serial.print("53L8A1 Simple Ranging demo application\n");
  Serial.print("--------------------------------------\n\n");

  Serial.print("Use the following keys to control application\n");
  Serial.print(" 'r' : change resolution\n");
  Serial.print(" 's' : enable signal and ambient\n");
  Serial.print(" 'c' : clear screen\n");
  Serial.print("\n");
}

void handle_cmd(uint8_t cmd)
{
  switch (cmd) {
    case 'r':
      toggle_resolution();
      clear_screen();
      break;

    case 's':
      toggle_signal_and_ambient();
      clear_screen();
      break;

    case 'c':
      clear_screen();
      break;

    default:
      break;
  }
}
```

**Expected Output from Above code:**
```
Entered Setup
Finished Setting up PWREN
Finished Setting up I2C Begin
Finished Setting up Sensor Begin
Finished Setting up Sensor Init
```

**Actual Output: from above code:**
```
Entered Setup
Finished Setting up PWREN
Finished Setting up I2C Begin
Finished Setting up Sensor Begin
```
**Finished Setting up Sensor Init** is missing.

Hence, I added a wrapper function in the **Documents\Arduino\libraries\STM32duino_VL53L8CX\src** folder to print the data from c file.

(PS: I have attached the STM32duino_VL53L8CX library folder with all my changes along with this issue for your reference)

**Changes made to the vl53l8cx_api.c file:**
```
#include "my_logging.h"
uint8_t vl53l8cx_init(
  VL53L8CX_Configuration    *p_dev)
{  my_log("Entered Init...");
  uint8_t tmp, status = VL53L8CX_STATUS_OK;
  uint8_t pipe_ctrl[] = {VL53L8CX_NB_TARGET_PER_ZONE, 0x00, 0x01, 0x00};
  uint32_t single_range = 0x01;
  uint32_t crc_checksum = 0x00;
  p_dev->default_xtalk = (uint8_t *)VL53L8CX_DEFAULT_XTALK;
  p_dev->default_configuration = (uint8_t *)VL53L8CX_DEFAULT_CONFIGURATION;
  p_dev->is_auto_stop_enabled = (uint8_t)0x0;
  my_log("Finished Init Configuration...");
  /* SW reboot sequence */
  status |= VL53L8CX_WrByte(&(p_dev->platform), 0x7fff, 0x00);
  status |= VL53L8CX_WrByte(&(p_dev->platform), 0x0009, 0x04);

  status |= VL53L8CX_WrByte(&(p_dev->platform), 0x000F, 0x40);
  status |= VL53L8CX_WrByte(&(p_dev->platform), 0x000A, 0x03);
  my_log("Finished Write Block...");
  status |= VL53L8CX_RdByte(&(p_dev->platform), 0x7FFF, &tmp);
  my_log("Finished Read Block...");
  status |= VL53L8CX_WrByte(&(p_dev->platform), 0x000C, 0x01);
...
....
```

**Expected Output from above lib code and ino code:**
```
Entered Setup
Finished Setting up PWREN
Finished Setting up I2C Begin
Finished Setting up Sensor Begin
Entered Init...
Finished Init Configuration...
Finished Write Block...
Finished Read Block...
```

**Actual Output from above lib code and ino dode:**
```
Entered Setup
Finished Setting up PWREN
Finished Setting up I2C Begin
Finished Setting up Sensor Begin
Entered Init...
Finished Init Configuration...
Finished Write Block...
```
**Finished Read Block...** is missing.

This is where my knowledge and experience ends. I was not able to debug more about the VL53L8CX_RdByte() func and its return value.

**Screenshots**
![image](https://github.com/user-attachments/assets/05df3eda-7221-4bb7-b624-e9ffb9a9d735)

**Desktop:**
 - OS: Windows 11
 - Arduino IDE version: 2.3.2
 - STM32 core version: 2.8.1
 - Library version: 2.0.0
 - Tools menu settings if not the default: Board-> Nucleo-64; Board Part Number-> Nucleo-F401RE; 
 - Upload method: SWD
![image](https://github.com/user-attachments/assets/b2bfdf00-4640-4339-a935-92b94c108724)

**Hardware:**
 - Board Name:Nucleo F401RE
 - Hardware Revision: Rev C
 - Extra hardware used if any: SATEL VL53L8 TOF sensor Breakout Board

**Additional Info**
Uploading the Library file as a zip file along with the wrapper function.
[STM32duino_VL53L8CX.zip](https://github.com/user-attachments/files/16487646/STM32duino_VL53L8CX.zip)

**Comment by cparata** [comment]
Hello @yaswanthsk04 ,
first of all, I would check the I2C connectivity. Could you run the following code with your hardware setup and see if the sensor is able to respond on the I2C bus?

```
// SPDX-FileCopyrightText: 2023 Carter Nelson for Adafruit Industries
//
// SPDX-License-Identifier: MIT
// --------------------------------------
// i2c_scanner
//
// Modified from https://playground.arduino.cc/Main/I2cScanner/
// --------------------------------------

#include <Wire.h>

// Set I2C bus to use: Wire, Wire1, etc.
#define DEV_I2C Wire
#define LPN_PIN A3
#define PWREN_PIN 11

void setup() {
  // Enable PWREN pin if present
  if (PWREN_PIN >= 0) {
    pinMode(PWREN_PIN, OUTPUT);
    digitalWrite(PWREN_PIN, HIGH);
    delay(10);
  }

  // Enable LPN pin if present
  if (LPN_PIN >= 0) {
    pinMode(LPN_PIN, OUTPUT);
    digitalWrite(LPN_PIN, HIGH);
    delay(10);
  }

  Serial.begin(9600);
  while (!Serial)
     delay(10);
  Serial.println("\nI2C Scanner");
  // Initialize I2C bus.
  DEV_I2C.begin();
  Serial.println("I2C started.");
}


void loop() {
  byte error, address;
  int nDevices;

  Serial.println("Scanning...");

  nDevices = 0;
  for(address = 1; address < 127; address++ )
  {
    // The i2c_scanner uses the return value of
    // the Write.endTransmisstion to see if
    // a device did acknowledge to the address.
    DEV_I2C.beginTransmission(address);
    error = DEV_I2C.endTransmission();

    if (error == 0)
    {
      Serial.print("I2C device found at address 0x");
      if (address<16)
        Serial.print("0");
      Serial.print(address,HEX);
      Serial.println("  !");

      nDevices++;
    }
    else if (error==4)
    {
      Serial.print("Unknown error at address 0x");
      if (address<16)
        Serial.print("0");
      Serial.println(address,HEX);
    }
  }
  if (nDevices == 0)
    Serial.println("No I2C devices found\n");
  else
    Serial.println("done\n");

  delay(5000);           // wait 5 seconds for next scan
}
```

**Comment by yaswanthsk04** [comment]
Hi @cparata:
Thank you for your reply.

The I2C Scanner output is follows:

```
Scanning...
No I2C devices Found
```

Thanks,
Yaswanthkumar

**Comment by cparata** [comment]
Hello @yaswanthsk04 ,
it means that there is an issue in your hardware setup. You should discover the ToF sensor with 0x29 I2C address. Until you don't get it, it does not make sense to run the VL53L8 library.
Please, check your wire connections.
Best Regards,
Carlo

**Comment by yaswanthsk04** [comment]
Hi @cparata:

Thank you for your reply.

The wire connections are given as per your instructions in Github page.

I doubted the connections too. So i used a multimeter to check the condution of each jumper wires and also used the multimeter to check the connections between the pins of vl53l8 sensor and nucleo board pins. The conduction is working fine on all pins.

are the connections in the github page right?

**Comment by cparata** [comment]
Hi @yaswanthsk04 ,
The connections in the guide should be fine. I checked the picture of your hardware setup and it seems fine to me even if I'm not able to see clearly all the connections. Please, be sure that SPI_I2C_N pin is connected to GND and check also the SCL pin that seems to me not well connected on the board.
Best Regards,
Carlo

**Comment by yaswanthsk04** [comment]
Hello @cparata:

Thank you for your reply.

based on the following Image from Application Note of ST Electronics on **How to connect VL53L8 to a nucleo board** I found that, the sensor must be powered using 5V pin instead of 3.3v pin

![image](https://github.com/user-attachments/assets/bfa00c3d-2f55-423e-aa79-db960cad7995)

Link to the Document: https://www.st.com/resource/en/application_note/an5945-how-to-connect-the-satelvl53l8-to-an-stm32-nucleo64-board-stmicroelectronics.pdf

Connecting the sensot to 5V pin works fine. I am able to find the I2C address as 0x29.

**Comment by cparata** [comment]
Oh great!
So, maybe now you can test the VL53L8 library and see if it works as well.
Best Regards,
Carlo

**Comment by yaswanthsk04** [comment]
Hello @cparata:

Yes, it is working now!!!

:)

![working](https://github.com/user-attachments/assets/bc592f1e-dced-4b2d-b1af-23802e7fdc62)

Did you connect it on 3.3v?

did it work for you in 3.3v?

**Comment by cparata** [comment]
Hello @yaswanthsk04 ,
Unfortunately, I don't have the possibility to check. I will update the README and the comments in the library to be aligned with the User Manual.
Thanks for have tested it.
Best Regards,
Carlo

**Comment by cparata** [comment]
I released a new version 2.0.1 fixing the comments in the README and in the comments inside the examples.
I close this issue.
Best Regards,
Carlo
